Automattic/wp-calypso
# Components: Migrate Dialog to TypeScript

#### Changes proposed in this Pull Request

* Migrate Dialog to TypeScript and make it a FunctionComponent

In part of making it a FunctionComponent I also extracted the `ButtonBar` concept into its own component for usage in the `Dialog`. The `buttons` prop was not previously documented so this should take some careful review to make sure I didn't miss anything from the original.

#### Testing instructions



* Go to `/domains/manage/:site/change-site-address/:site` and ensure changing the site address works.
* Go to the language picker and ensure changing languages works

These two scenarios cover the two types of `buttons` arguments: `React.ReactElement`s and the `BaseButton` style object configuration buttons. Feel free to test any other dialogs in the app for good measure.
